shakenfist/daemons/cluster.py

```diff
@@ -53,21 +56,116 @@ def _await_election(self):
+                    if (obj.state.value == dbo.STATE_DELETED and
+                            time.time() - obj.state.update_time > config.CLEANER_DELAY):
+                        LOG.with_fields({
+                            objtype: obj.uuid
+                        }).info('Hard deleting')
+                        obj.hard_delete()
+                except exceptions.BadObjectVersion:
                     LOG.with_fields({
                         objtype: obj.uuid
-                    }).info('Hard deleting')
-                    obj.hard_delete()
+                    }).warning('Could not load object for hard delete, bad version')
 
         # Prune artifacts which might have too many versions
         for a in artifact.Artifacts([]):
             a.delete_old_versions()
 
+        # Prune over replicated blobs
+        overreplicated = {}
+        underreplicated = []
+        nodes_by_disk_ascending = sorted(
+            nodes_by_free_disk_descending(), reverse=True)
+        low_disk_nodes = nodes_by_free_disk_descending(
+            minimum=0, maximum=config.MINIMUM_FREE_DISK)
+        absent_nodes = [n.fqdn for n in Nodes([node_inactive_states_filter])]
+
+        with etcd.ThreadLocalReadOnlyCache():
+            for b in blob.Blobs([active_states_filter]):
+                locations = b.locations
+                ignored_locations = []
+                for n in absent_nodes:
+                    if n in locations:
+                        locations.remove(n)
+                        ignored_locations.append(n)
+
+                if ignored_locations:
+                    LOG.with_fields({
+                        'blob': b,
+                        'ignored_locations': ignored_locations
+                    }).info('Ignored some blob locations as nodes are absent')
+
+                delta = len(locations) - config.BLOB_REPLICATION_FACTOR
+                if delta > 0:
+                    excess_locations = b.locations
+                    in_use_locations = []
+
+                    for instance_uuid in b.instances:
+                        i = instance.Instance.from_db(instance_uuid)
+                        node = i.placement.get('node')
+                        if node in excess_locations:
+                            excess_locations.remove(node)
+                            in_use_locations.append(node)
+
+                    if len(in_use_locations) >= config.BLOB_REPLICATION_FACTOR:
+                        # We have enough locations in use to meet the replication
+                        # factor, remove everything else
+                        overreplicated[b.uuid] = excess_locations
+                    else:
+                        # We need to keep some idle locations to maintain our
+                        # replication factor. Prefer to remove the fullest first.
+                        overreplicated[b.uuid] = []
+                        target = config.BLOB_REPLICATION_FACTOR - \
```

Comment:
PEP8 prefers brackets for line continuation

Reply:
As do I. I think this was my editor. I will fix.